<a href="https://colab.research.google.com/github/NikhilGeorge01/ML-Prac/blob/main/EnsembleLearning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from sklearn.datasets import load_iris
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
data = load_iris()

In [ ]:
dt = DecisionTreeClassifier(max_depth=3, random_state=40)
rf = RandomForestClassifier(n_estimators=3, random_state=40)
X = data.data
y = data.target

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size = 0.2, random_state=40)


In [ ]:
rf.fit(X_train,y_train)


RandomForestClassifier(n_estimators=3, random_state=40)

In [ ]:
dt.fit(X_train,y_train)


DecisionTreeClassifier(max_depth=3, random_state=40)

In [ ]:
rf_pred = rf.predict(X_test)
dt_pred = dt.predict(X_test)

In [ ]:
print("FOR DECISION TREE")
print(f"Accuracy is {accuracy_score(y_test,dt_pred)} Precision is {precision_score(y_test,dt_pred, average='macro')} Recall is {recall_score(y_test,dt_pred, average='macro')} F1 score is {f1_score(y_test,dt_pred, average='macro')}")

FOR DECISION TREE
Accuracy is 0.9666666666666667 Precision is 0.9743589743589745 Recall is 0.9666666666666667 F1 score is 0.9691228070175438


In [ ]:
print("FOR RANDOM FOREST")
print(f"Accuracy is {accuracy_score(y_test,rf_pred)} Precision is {precision_score(y_test,rf_pred, average='macro')} Recall is {recall_score(y_test,rf_pred, average='macro')} F1 score is {f1_score(y_test,rf_pred, average='macro')}")

FOR RANDOM FOREST
Accuracy is 1.0 Precision is 1.0 Recall is 1.0 F1 score is 1.0


FOR DECISION TREE WE GET VALUES WHILE TESTING AS Accuracy is 0.9666666666666667 Precision is 0.9743589743589745 Recall is 0.9666666666666667 F1 score is 0.9691228070175438, and for random forest classifier as we increse number of estimators(dec trees) we notice that accuracy keeps increaseing

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier

In [ ]:
log_clf = LogisticRegression(max_iter=1000)
rf_clf = RandomForestClassifier(n_estimators=100, random_state=40)
knn_clf = KNeighborsClassifier(n_neighbors=5)
from sklearn.ensemble import VotingClassifier

In [ ]:
voting_hard = VotingClassifier(
    estimators=[('lr', log_clf), ('rf', rf_clf), ('knn', knn_clf)],
    voting='hard'
)

voting_hard.fit(X_train, y_train)
y_pred = voting_hard.predict(X_test)
print("Hard Voting Accuracy:", accuracy_score(y_test, y_pred))


Hard Voting Accuracy: 1.0


In [ ]:
voting_soft = VotingClassifier(
    estimators=[('lr', log_clf), ('rf', rf_clf), ('knn', knn_clf)],
    voting='soft'
)

voting_soft.fit(X_train, y_train)
y_pred = voting_soft.predict(X_test)
print("Soft Voting Accuracy:", accuracy_score(y_test, y_pred))


Soft Voting Accuracy: 1.0


In [ ]:
voting_weighted = VotingClassifier(
    estimators=[('lr', log_clf), ('rf', rf_clf), ('knn', knn_clf)],
    voting='soft',
    weights=[1, 3, 2]  # higher weight to Random Forest
)

voting_weighted.fit(X_train, y_train)
y_pred = voting_weighted.predict(X_test)
print("Weighted Soft Voting Accuracy:", accuracy_score(y_test, y_pred))


Weighted Soft Voting Accuracy: 1.0


In [2]:
from sklearn.datasets import load_diabetes
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

data_reg = load_diabetes()
X_reg = data_reg.data
y_reg = data_reg.target

X_train_reg, X_test_reg, y_train_reg, y_test_reg = train_test_split(X_reg, y_reg, test_size=0.2, random_state=40)

rf_reg = RandomForestRegressor(n_estimators=100, random_state=40, oob_score=True)
rf_reg.fit(X_train_reg, y_train_reg)

print("OOB Score:", rf_reg.oob_score_)

y_pred_reg = rf_reg.predict(X_test_reg)
print("MSE on Test Set:", mean_squared_error(y_test_reg, y_pred_reg))

OOB Score: 0.47143555957471794
MSE on Test Set: 4540.1692617977515


In [3]:
from sklearn.datasets import load_breast_cancer
from sklearn.ensemble import AdaBoostClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

data_bc = load_breast_cancer()
X_bc = data_bc.data
y_bc = data_bc.target

X_train_bc, X_test_bc, y_train_bc, y_test_bc = train_test_split(X_bc, y_bc, test_size=0.2, random_state=40)

adaboost = AdaBoostClassifier(n_estimators=100, random_state=40)
adaboost.fit(X_train_bc, y_train_bc)

y_pred_bc = adaboost.predict(X_test_bc)
print("AdaBoost Accuracy:", accuracy_score(y_test_bc, y_pred_bc))

AdaBoost Accuracy: 0.956140350877193
